### ISSUE TITLE
Holding Escape forever pauses the game

### ISSUE THREAD

**Issue by LunaticEdit** [OpenRA/OpenRA#7941]
In a single player game, start it up, and at whatever point, hold down ESC for about 5 seconds. Game is now forever paused. In a single player game, start it up, and at whatever point, hold down ESC for about 5 seconds. Game is now forever paused. Bleed 70c9bca
##

**Comment by LunaticEdit** [comment]
Just some additional info. It appears the pause state flags somehow get flipped at some point. It smells of a race condition that may have something to do with the fact it sets a local pause state, AND sends the message to the server, which also gets processed as well (in local mode).

**Comment by RichSuther** [comment]
Investigated this one a bit, but I see this as problem in two parts.

Firstly to confirm it is indeed being cached on single player

In MenuButtonsChromeLogic.cs
```
		void OpenMenuPanel(MenuButtonWidget button, WidgetArgs widgetArgs = null)
		{
			disableSystemButtons = true;
			var cachedPause = world.PredictedPaused;

			if (button.HideIngameUI)
				worldRoot.IsVisible = () => false;

			if (button.Pause && world.LobbyInfo.IsSinglePlayer)
				world.SetPauseState(true);

			widgetArgs = widgetArgs ?? new WidgetArgs();
			widgetArgs.Add("onExit", () =>
			{
				if (button.HideIngameUI)
					worldRoot.IsVisible = () => true;

				if (button.Pause && world.LobbyInfo.IsSinglePlayer)
					world.SetPauseState(cachedPause);

				menuRoot.RemoveChild(currentWidget);
				disableSystemButtons = false;
			});

			currentWidget = Game.LoadWidget(world, button.MenuContainer, menuRoot, widgetArgs);
		}

```
Inside the World class there is a method (SetLocalPausedState - never used?) that skips using the IssueOrder and sets Paused and PredictedPause, this does indeed solve the issue just using SetLocalPausedState(true/false) accordingly.

But since I'm a newbie to OpenRA development it would be nice if someone more experienced could double check that is indeed an acceptable way to fix this, as I know this relates to communication between threads.


Also Secondly It should not be possible to hold down a keyboard key and for it to fire every tick of the keyboard this same behaviour is seen throughout the sidebar (Main menu, Beacon placement etc.) I think that in itself is a bug.

**Comment by abcdefg30** [comment]
> Inside the World class there is a method (SetLocalPausedState - never used?) that skips using the IssueOrder and sets Paused and PredictedPause, this does indeed solve the issue just using SetLocalPausedState(true/false) accordingly.

I think it is a good idea to "just use SetLocalPausedState(true/false)" and remove `cachedPause` (or is there a good reason why this is needed?). The only case I can think about is manually pausing before pressing `esc`.

**Comment by RichSuther** [comment]
I asked on IRC yesterday, @pchote indicated this would be problematic behind the scenes.
```
[17:18:39]	<pchote> there's asynchronous code going on there
[17:18:50]	<pchote> its not as simple as just using the world state
[17:22:56]	<pchote> its there to work around a lag that would otherwise exist
[17:23:13]	<pchote> it takes ~100ms (even on skirmish) for an order to roundtrip
[17:23:33]	<pchote> that's long enough to be annoying, waiting for the menu to open
[17:23:37]	<pchote> its significantly worse on mp games 
[17:28:56]	<pchote> should be straightforward to add a check for when the state is dirty, and behave properly
[17:29:03]	<pchote> if predicted state != world state
[17:29:29]	<pchote> not sure off the top of my head what "behave properly" involves in this case - without introducing unwanted sideeffects 
[17:33:10]	<pchote> because you can then create a situation where the menu is open locally, but the world thinks its closed
[17:34:10]	<pchote> I suspect that the better fix is to handle key repeats better 
```
I agree that key repeats should be handled better maybe make that a separate issue?, but it would be nice to ensure the state is guarded from being locked as paused in either case as it may pop up again.

**Comment by rob-v** [comment]
Fixed by #13323 thanks to recent KeyInput.IsRepeat field.
